derp/derp_server.go

```diff
@@ -1942,6 +2085,113 @@ func (s *Server) ServeDebugTraffic(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// takeFlowsSnapshot returns a map of copies of all the flowStats for
+// this server. The output should be suitable for comparing the stats
+// for a specific flow at multiple points in time so that we can
+// calculate the bandwidth used.
+func (s *Server) takeFlowsSnapshot() map[*sclient]map[key.NodePublic]map[key.NodePublic]*flowStats {
+	s.debugLogf("taking flows snapshot")
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	snap := maps.Clone(s.flows)
+	for c := range s.flows {
+		snap[c] = maps.Clone(s.flows[c])
+		for dstKey := range s.flows[c] {
+			snap[c][dstKey] = maps.Clone(s.flows[c][dstKey])
+			for srcOrFwd := range s.flows[c][dstKey] {
+				f := flowStats{}
+				f = *s.flows[c][dstKey][srcOrFwd]
+				snap[c][dstKey][srcOrFwd] = &f
+			}
+		}
+	}
+	return snap
+}
+
+// ServeFlows prints a list of recently active data flows through
+// this server, including the bandwidth used by each over the last
+// second.
+func (s *Server) ServeFlows(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "current time: %s\n", time.Now().Format("2006-01-02 15:04:05"))
+
+	lastTime := time.Now()
+	lastSnap := s.takeFlowsSnapshot()
+	// Sleep long enough to get some kind of useful bandwidth estimate
+	time.Sleep(1 * time.Second)
```

Comment:
Maybe make this a URL parameter? It can default to 1 second, if that's good. It feels low, though.

Better would probably be a background snapshotter every 30 seconds or something, then this handler doesn't need to block at all.